manifests/server/config.pp

```diff
@@ -174,6 +174,18 @@
         Exec['puppet_server_config-create_ssl_dir'],
       ],
     }
+
+    # In Puppet 7 the cadir was changed from $ssldir/ca to $puppetserver_dir/ca
+    # This migrates the directory if it was in the old location
+    # The migration command leaves a symlink in place
+    if versioncmp($puppet::server::real_puppetserver_version, '7.0') > 0 {
+      exec { 'migrate Puppetserver cadir':
+        command => "${puppet::puppetserver_cmd} ca migrate",
```

Comment:
This returns:
```
Notice: /Stage[main]/Puppet::Server::Config/Exec[migrate Puppetserver cadir]/returns: Puppetserver service is running. Please stop it before attempting to run this command.
Error: '/opt/puppetlabs/bin/puppetserver ca migrate' returned 1 instead of one of [0]
Error: /Stage[main]/Puppet::Server::Config/Exec[migrate Puppetserver cadir]/returns: change from 'notrun' to ['0'] failed: '/opt/puppetlabs/bin/puppetserver ca migrate' returned 1 instead of one of [0]
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: The cadir is currently configured to be inside the /etc/puppetlabs/puppet/ssl directory. This config setting and the directory location will not be used in a future version of puppet. Please run the puppetserver ca tool to migrate out from the puppet confdir to the /etc/puppetlabs/puppetserver/ca directory. Use `puppetserver ca migrate --help` for more info.
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Error:
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_crt.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_crl.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/infra_crl.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/certs/centos8.wisse.example.com.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/certs/ca.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/crl.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_pub.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/inventory.txt'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/infra_inventory.txt'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/infra_serials'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/serial'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/signed/centos8.wisse.example.com.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/root_key.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Existing file at '/etc/puppetlabs/puppet/ssl/ca/ca_key.pem'
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: If you would really like to replace your CA, please delete the existing files first.
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: Note that any certificates that were issued by this CA will become invalid if you
Notice: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: replace it!
Error: '/opt/puppetlabs/bin/puppetserver ca setup' returned 1 instead of one of [0]
Error: /Stage[main]/Puppet::Server::Config/Exec[puppet_server_config-generate_ca_cert]/returns: change from 'notrun' to ['0'] failed: '/opt/puppetlabs/bin/puppetserver ca setup' returned 1 instead of one of [0]
Notice: /Stage[main]/Puppet::Server::Service/Service[puppetserver]: Dependency Exec[puppet_server_config-generate_ca_cert] has failures: true
Notice: /Stage[main]/Puppet::Server::Service/Service[puppetserver]: Dependency Exec[migrate Puppetserver cadir] has failures: true
```

Tomorrow I'll hopefully be able to come up with a solution.

Reply:
I've now added an item to README. For the installer I'll write a hook.